docs/design_proposals/ko-builder.md

```diff
@@ -0,0 +1,240 @@
+  [fast](https://cloud.google.com/blog/topics/developers-practitioners/ship-your-go-applications-faster-cloud-run-ko).
+
+- _share with other developers_: no additional tools are required to
+  `skaffold run` with the ko builder, not even Docker. Though if we don't embed
+  ko in Skaffold, ko will be a tool that all developers in a team would have to
+  install.
+
+- works great as a _CI/CD building block_: when using the ko builder, pipeline
+  steps can run using the default Skaffold container image, without
+  installing additional tools or keeping toolchain versions in sync across
+  local development and CI/CD.
+
+## Design
+
+Adding the ko builder requires making config changes to the Skaffold schema.
+
+1.  Add an entry to BuilderType enum in `proto/enums/enums.proto`:
+
+    ```proto
+        // ko Builder
+        KO = 7;
+    ```
+
+2.  Add a `KoArtifact` type:
+
+    ```go
+    type KoArtifact struct {
+    	// Annotations are key-value string pairs to add to the image manifest.
+    	// Also known as `LABEL` in `Dockerfile`s.
+    	// Ref: https://github.com/opencontainers/image-spec/blob/master/annotations.md
+    	Annotations map[string]string `yaml:"annotations,omitempty"`
+
+    	// BaseImage overrides the default ko base image.
+    	// Corresponds to, and overrides, the `defaultBaseImage` in `.ko.yaml`.
+    	BaseImage string `yaml:"fromImage,omitempty"`
+
+    	// Env are environment variables, in the `key=value` form, passed to the build.
+    	// For example: `CGO_ENABLED=1`.
+    	Env []string `yaml:"env,omitempty"`
+
+    	// Platforms is the list of platforms to build images for. Each platform
+    	// is of the format `os/arch[/variant]`, e.g., `linux/amd64`.
+    	// By default, the ko builder builds for `all` platforms supported by the
+    	// base image.
+    	Platforms []string `yaml:"platforms,omitempty"`
+
+    	// SourceDateEpoch is the `created` time of the container image.
+    	// Specify as the number of seconds since January 1st 1970, 00:00 UTC.
+    	// You can override this value by setting the `SOURCE_DATE_EPOCH`
+    	// environment variable.
+    	SourceDateEpoch unit64 `yaml:"sourceDateEpoch,omitempty"`
+    }
+    ```
+
+3.  Add a `KoArtifact` field to the `ArtifactType` struct:
+
+    ```go
+    type ArtifactType struct {
+    	// KoArtifact builds images using [ko](https://github.com/google/ko).
+    	KoArtifact *KoArtifact `yaml:"ko,omitempty" yamltags:"oneOf=artifact"`
+    }
+    ```
+
+Example basic config, this will be sufficient for many users:
+
+```yaml
+apiVersion: skaffold/v2beta14
+kind: Config
+build:
+  artifacts:
+  - image: ko://example.com/helloworld
+    ko: {}
+```
+
+The value of the `image` field is the Go import path of the app entry point,
+[prefixed by `ko://`](https://github.com/google/ko/pull/58).
+
+ko requires setting a
+[`KO_DOCKER_REPO`](https://github.com/google/ko#choose-destination)
+environment variable. This determines where container images are pushed.
```

Comment:
Ko publishes an image as `registry:/blah/app-MD5`. Skaffold relies on pushing the image as per the tag policy. I am not sure how we do this for `jib`, but we could look into adding a `dest` flag for `ko`

Reply:
Currently, when using ko via the custom builder functionality, image names become `[repository]/[importpath]`. This matches what ko produces when running `ko publish` with the `--preserve-import-paths` flag. This is also how Knative and Tekton publish release images, so it would work well as a default.

ko provides other ways to customize the image name (`--bare` and `--base-import-paths`). Whether we'd want support these through Skaffold, or add a `dest` config option as you suggest is an open question.

Reply:
sounds good.